### ISSUE TITLE
Add more explanation on multithreaded graph building of Autograd

### ISSUE THREAD

**Issue by ifsheldon** [pytorch/pytorch#67887]
## 📚 Documentation

In the [Autograd Mechanics - multithreaded-autograd](https://pytorch.org/docs/stable/notes/autograd.html#multithreaded-autograd), the doc mentions that there're risks to run backward() or grad() in multiple threads. However, it does not mention how Autograd will work in some simpler cases. For example, a mapping function like:
```python
def map(tensor):
    if tensor.shape[0] > 10:
        return tensor.sum()
    else:
        return 2 * tensor

from multiprocessing import pool
thread_pool = pool.ThreadPool()
some_tensors = [......]
tensors = pool.map(some_tensors)
loss = torch.zeros(1)
for t in tensors:
    loss += (t**2).sum()

loss.backward()
```
Will this always work? I think as long as there are no inplace operations (including .backward() or .grad()), it is safe, but can you please confirm that and update this page a bit?

**If so**, can you please update [pytorch multiprocessing](https://pytorch.org/docs/stable/multiprocessing.html) as well? I think the `torch.multiprocessing` is just a subset of Python's `multiprocessing`, since Python's one has `ThreadPool` while `torch.multiprocessing` does not. 

When I tried to do the above simple mapping using `torch.multiprocessing.Pool`, an error occurs, saying that Autograd cannot cross process boundary. It's understandable, but I think you should modify this line
> The API is 100% compatible with the original module

and suggest users to use the original one if they only want to do some mapping that cannot be expressed as batched operations but can be safely executed with Autograd concurrently with multiple threads.

**Comment by albanD** [comment]
I am not sure follow you problem:
- Both multiprocessing pool will fail if they try to use multiple processes (in your example able you compare a thread pool with a process pool no?)
- The threading will always work.

**Comment by ifsheldon** [comment]
>Both multiprocessing pool will fail if they try to use multiple processes (in your example able you compare a thread pool with a process pool no?)

Yes, I was comparing `torch.multiprocessing` with the original one. I take `ThreadPool` as an example for showing `torch.multiprocessing` is only a subset of the original one, not "100% compatible with the original module" as claimed, since the original one has `ThreadPool`. 

And since the `torch.multiprocessing` is only a subset, it'd better to suggest people to use the original one when they want Autograd with concurrency, and explicitly state when Autograd will fail and when it will not.

> The threading will always work.

As written in this issue earlier, in the [Autograd Mechanics - multithreaded-autograd](https://pytorch.org/docs/stable/notes/autograd.html#multithreaded-autograd), the doc mentions that there're risks to run backward() or grad() in multiple threads. Therefore, the threading will not always work. By "work" I mean Autograd gives calculates correct gradients. 

|      | With Concurrent Inplace Operations(e.g. .backward(), .grad(), +=) |            Without Concurrent Inplace Operations             |
| :--: | :----------------------------------------------------------: | :----------------------------------------------------------: |
| GPU  | Not always work, may give wrong gradients (due to potential data races) | Always works? Can we submit multiple tensor operation concurrently? |
| CPU  | Not always work, may give wrong gradients (due to potential data races) |                Always works with ThreadPool?                 |

We have already know the left part of the table, but not the right part. So, in this issue, I was trying to urge you to clarify the right part and update PyTorch's documentation accordingly.

**Comment by albanD** [comment]
The idea for multithreading with autograd is the same as any other multithreaded operation in python.
You can read from multiple threads without any worry. But if multiple threads modify an object inplace, you will have to do some locking to ensure consistency of the result.
In this context, it is important so know that `.backward()` is an inplace op if any of the `.grad` is already specified. But `autograd.grad()` is never inplace.

In general:
- You can do any inplace you want during the forward but need to do proper locking if objects being modified are used in different threads.
- Backward may also be inplace so you might want to do locking around it if different thread's graphs share the same leaf variables.
- grad() is never inplace so it is always safe to call from multiple threads (even on the same graph!)
- the backward pass it self won't do any inplace into shared objects so that is safe. But if you use custom autograd.Function that you wrote, you want to make sure that these are also thread safe.


For multiprocessing, after checking, we do `from multiprocessing import *` in our multiprocessing package so I am surprised that you see missing objects.

**Comment by ifsheldon** [comment]
Thanks for the confirmation. Then please update the documentation a bit accordingly to solidify the confirmation explicitly.

> For multiprocessing, after checking, we do from multiprocessing import * in our multiprocessing package so I am surprised that you see missing objects.

After seeing your explanation, I check the code as well, and indeed you do `from multiprocessing import *` in its `__init__.py`, so now it surprises me too.

Here is the Pytorch version I've been using.

```
# Name                    Version                   Build  Channel
pytorch                   1.10.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
```

And I just simply run this code to test `torch.multiprocessing`.
```python
from torch import multiprocessing as mp
import torch


def test_multithreading():
    a = mp.pool.ThreadPool()  # Not working
    a.close()
    a = mp.ThreadPool()  # Not working
    a.close()
```
Did I do anything wrong?

**Comment by albanD** [comment]
> Did I do anything wrong?

I don't think so, but that doesn't work with the plain multiprocessing module either:
```python
import multiprocessing
print(multiprocessing.pool)
```
will return
```
AttributeError: module 'multiprocessing' has no attribute 'pool'
```

You will have to explicitly import it.
That being said, the `pool` there only has the plain `Pool`. We can expose back the thread pool as well there if needed.

**Comment by ifsheldon** [comment]
OK, there're some problems with module importing I think.

But I can use ThreadPool with `from multiprocessing import pool`, but not with `from torch.multiprocessing import pool`.
```
# from multiprocessing import pool  # this works
from torch.multiprocessing import pool  # not working


def test_multithreading():
    a = pool.ThreadPool()
    print(a)
    a.close()
```

**Comment by albanD** [comment]
We should:
- properly expose the `pool` submodule in `torch.multiprocessing`
- Clarify the multithreaded expectation with autograd (and Tensor objects in general)
